util/network.go

```diff
@@ -15,16 +15,26 @@ import (
 // CreateListener crates socket listener based on DSN definition.
 func CreateListener(address string) (net.Listener, error) {
 	dsn := strings.Split(address, "://")
-	if len(dsn) != 2 {
-		return nil, errors.New("invalid DSN (tcp://:6001, unix://file.sock)")
-	}
 
-	if dsn[0] != "unix" && dsn[0] != "tcp" {
-		return nil, errors.New("invalid Protocol (tcp://:6001, unix://file.sock)")
+	var protocol string
+	var param    string
+
+	if len(dsn) == 1 {
+		protocol = "tcp"
+		param = dsn[0]
+	} else if len(dsn) == 2 {
+		if dsn[0] != "unix" && dsn[0] != "tcp" {
+			return nil, errors.New("invalid Protocol ([tcp://]:6001, unix://file.sock)")
+		} else {
+			protocol = dsn[0]
+			param = dsn[1]
+		}
+	} else {
+		return nil, errors.New("invalid DSN ([tcp://]:6001, unix://file.sock)")
```

Comment:
I do not fully understand the parsing path here. If you assume, that `if len(dsn) == 1`, you set `protocol` variable to `"tcp"`, but at the same time you checking `if protocol == "unix" && fileExists(param)` which is redundant in this execution path because in case of `tcp` you already know, that protocol is not an `unix`.

Reply:
What I implemented is:

 - if no prefix is presented, default to TCP (e.g. `127.0.0.1:8000`)
 - if `tcp://` is presented, assume TCP (e.g. `[tcp://]127.0.0.1:8000`)
 - if `unix://` is presented, assume UNIX (e.g. `unix:///path/to/unix.socket`)
 - if a prefix is presented, but neither `unix://` nor `tcp://`, error out.

This is meant to avoid introducing a breaking change. Originally, this field accepts only `ip:port`, but now both `ip:port` and `tcp://ip:port` are accepted, which kept `ip:port` acceptable.

maybe the variable naming is an issue, maybe the following change can be better?

 - `protocol` => `socket_type`
 - `param` => `address`

Reply:
I suggest you rewrite this function in the following way:
```go
// CreateListener crates socket listener based on DSN definition.
func CreateListener(address string) (net.Listener, error) {
	dsn := strings.Split(address, "://")

	switch len(dsn) {
	case 1:
		// assume, that there is no prefix here [127.0.0.1:8000]
		return createTCPListener(dsn[0])
	case 2:
                // we got two part here, first part is the transport, second - address
		// [tcp://127.0.0.1:8000] OR [unix:///path/to/unix.socket] OR [error://path]
		// where error is wrong transport name
		switch dsn[0] {
		case "unix":
			// check of file exist. If exist, unlink
			if fileExists(dsn[1]) {
				err := syscall.Unlink(dsn[1])
				if err != nil {
					return nil, fmt.Errorf("error during the unlink syscall: error %v", err)
				}
			}
			return net.Listen(dsn[0], dsn[1])
		case "tcp":
			return createTCPListener(dsn[1])
			// not an tcp or unix
		default:
			return nil, fmt.Errorf("invalid Protocol ([tcp://]:6001, unix://file.sock), address: %s", address)
		}
		// wrong number of split parts
	default:
		return nil, fmt.Errorf("wrong number of parsed protocol parts, address: %s", address)
	}
}

func createTCPListener(addr string) (net.Listener, error) {
	cfg := tcplisten.Config{
		ReusePort:   true,
		DeferAccept: true,
		FastOpen:    true,
		Backlog:     0,
	}
	listener, err := cfg.NewListener("tcp4", addr)
	if err != nil {
		return nil, err
	}
	return listener, nil
}
```
It's simpler, without new variables, without a bunch of `if-else` (which makes readability harder). I wrote this in the browser, so, it may contain an issue :)

Reply:
Addresses in 18fea30